# 24_문자열 응용하기

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#문자열-조작하기" data-toc-modified-id="문자열-조작하기-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>문자열 조작하기</a></span><ul class="toc-item"><li><span><a href="#문자열-바꾸기" data-toc-modified-id="문자열-바꾸기-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>문자열 바꾸기</a></span></li><li><span><a href="#문자-바꾸기" data-toc-modified-id="문자-바꾸기-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>문자 바꾸기</a></span></li><li><span><a href="#문자열-분리하기" data-toc-modified-id="문자열-분리하기-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>문자열 분리하기</a></span></li><li><span><a href="#구분자-문자열과-문자열-리스트-연결하기" data-toc-modified-id="구분자-문자열과-문자열-리스트-연결하기-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>구분자 문자열과 문자열 리스트 연결하기</a></span></li><li><span><a href="#소문자를-대문자로-바꾸기" data-toc-modified-id="소문자를-대문자로-바꾸기-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>소문자를 대문자로 바꾸기</a></span></li><li><span><a href="#대문자를-소문자로-바꾸기" data-toc-modified-id="대문자를-소문자로-바꾸기-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>대문자를 소문자로 바꾸기</a></span></li><li><span><a href="#왼쪽-공백-삭제하기" data-toc-modified-id="왼쪽-공백-삭제하기-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>왼쪽 공백 삭제하기</a></span></li><li><span><a href="#오른쪽-공백-삭제하기" data-toc-modified-id="오른쪽-공백-삭제하기-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>오른쪽 공백 삭제하기</a></span></li><li><span><a href="#양쪽-공백-삭제하기" data-toc-modified-id="양쪽-공백-삭제하기-1.9"><span class="toc-item-num">1.9&nbsp;&nbsp;</span>양쪽 공백 삭제하기</a></span></li><li><span><a href="#왼쪽의-특정-문자-삭제하기" data-toc-modified-id="왼쪽의-특정-문자-삭제하기-1.10"><span class="toc-item-num">1.10&nbsp;&nbsp;</span>왼쪽의 특정 문자 삭제하기</a></span></li><li><span><a href="#오른쪽-특정-문자-삭제하기" data-toc-modified-id="오른쪽-특정-문자-삭제하기-1.11"><span class="toc-item-num">1.11&nbsp;&nbsp;</span>오른쪽 특정 문자 삭제하기</a></span></li><li><span><a href="#양쪽의-특정-문자-삭제하기" data-toc-modified-id="양쪽의-특정-문자-삭제하기-1.12"><span class="toc-item-num">1.12&nbsp;&nbsp;</span>양쪽의 특정 문자 삭제하기</a></span></li><li><span><a href="#문자열을-왼쪽-정렬하기" data-toc-modified-id="문자열을-왼쪽-정렬하기-1.13"><span class="toc-item-num">1.13&nbsp;&nbsp;</span>문자열을 왼쪽 정렬하기</a></span></li><li><span><a href="#문자열을-오른쪽-정렬하기" data-toc-modified-id="문자열을-오른쪽-정렬하기-1.14"><span class="toc-item-num">1.14&nbsp;&nbsp;</span>문자열을 오른쪽 정렬하기</a></span></li><li><span><a href="#문자열을-가운데-정렬하기" data-toc-modified-id="문자열을-가운데-정렬하기-1.15"><span class="toc-item-num">1.15&nbsp;&nbsp;</span>문자열을 가운데 정렬하기</a></span></li><li><span><a href="#메서드-체이닝" data-toc-modified-id="메서드-체이닝-1.16"><span class="toc-item-num">1.16&nbsp;&nbsp;</span>메서드 체이닝</a></span></li><li><span><a href="#문자열-왼쪽에-0-채우기" data-toc-modified-id="문자열-왼쪽에-0-채우기-1.17"><span class="toc-item-num">1.17&nbsp;&nbsp;</span>문자열 왼쪽에 0 채우기</a></span></li><li><span><a href="#문자열-위치-찾기" data-toc-modified-id="문자열-위치-찾기-1.18"><span class="toc-item-num">1.18&nbsp;&nbsp;</span>문자열 위치 찾기</a></span></li><li><span><a href="#오른쪽에서부터-문자열-위치-찾기" data-toc-modified-id="오른쪽에서부터-문자열-위치-찾기-1.19"><span class="toc-item-num">1.19&nbsp;&nbsp;</span>오른쪽에서부터 문자열 위치 찾기</a></span></li><li><span><a href="#문자열-위치-찾기" data-toc-modified-id="문자열-위치-찾기-1.20"><span class="toc-item-num">1.20&nbsp;&nbsp;</span>문자열 위치 찾기</a></span></li><li><span><a href="#오른쪽에서부터-문자열-위치-찾기" data-toc-modified-id="오른쪽에서부터-문자열-위치-찾기-1.21"><span class="toc-item-num">1.21&nbsp;&nbsp;</span>오른쪽에서부터 문자열 위치 찾기</a></span></li><li><span><a href="#문자열-개수-세기" data-toc-modified-id="문자열-개수-세기-1.22"><span class="toc-item-num">1.22&nbsp;&nbsp;</span>문자열 개수 세기</a></span></li></ul></li><li><span><a href="#문자열-서식-지정자와-포매팅-사용하기" data-toc-modified-id="문자열-서식-지정자와-포매팅-사용하기-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>문자열 서식 지정자와 포매팅 사용하기</a></span><ul class="toc-item"><li><span><a href="#서식-지정자로-문자열-넣기" data-toc-modified-id="서식-지정자로-문자열-넣기-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>서식 지정자로 문자열 넣기</a></span></li><li><span><a href="#서식-지정자로-숫자-넣기" data-toc-modified-id="서식-지정자로-숫자-넣기-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>서식 지정자로 숫자 넣기</a></span></li><li><span><a href="#서식-지정자로-소수점-표현하기" data-toc-modified-id="서식-지정자로-소수점-표현하기-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>서식 지정자로 소수점 표현하기</a></span></li><li><span><a href="#서식-지정자로-문자열-정리하기" data-toc-modified-id="서식-지정자로-문자열-정리하기-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>서식 지정자로 문자열 정리하기</a></span></li><li><span><a href="#서식-지정자로-문자열-안에-값-여러-개-넣기" data-toc-modified-id="서식-지정자로-문자열-안에-값-여러-개-넣기-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>서식 지정자로 문자열 안에 값 여러 개 넣기</a></span></li><li><span><a href="#format-메서드-사용하기" data-toc-modified-id="format-메서드-사용하기-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>format 메서드 사용하기</a></span></li><li><span><a href="#format-메서드로-값을-여러-개-넣기" data-toc-modified-id="format-메서드로-값을-여러-개-넣기-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>format 메서드로 값을 여러 개 넣기</a></span></li><li><span><a href="#format-메서드로-같은-값을-여러-개-넣기" data-toc-modified-id="format-메서드로-같은-값을-여러-개-넣기-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>format 메서드로 같은 값을 여러 개 넣기</a></span></li><li><span><a href="#format-메서드에서-인덱스-생략하기" data-toc-modified-id="format-메서드에서-인덱스-생략하기-2.9"><span class="toc-item-num">2.9&nbsp;&nbsp;</span>format 메서드에서 인덱스 생략하기</a></span></li><li><span><a href="#format-메서드에서-인덱스-대신-이름-지정하기" data-toc-modified-id="format-메서드에서-인덱스-대신-이름-지정하기-2.10"><span class="toc-item-num">2.10&nbsp;&nbsp;</span>format 메서드에서 인덱스 대신 이름 지정하기</a></span></li><li><span><a href="#문자열-포매팅에-변수를-그대로-사용하기" data-toc-modified-id="문자열-포매팅에-변수를-그대로-사용하기-2.11"><span class="toc-item-num">2.11&nbsp;&nbsp;</span>문자열 포매팅에 변수를 그대로 사용하기</a></span></li><li><span><a href="#format-메서드로-문자열-정렬하기" data-toc-modified-id="format-메서드로-문자열-정렬하기-2.12"><span class="toc-item-num">2.12&nbsp;&nbsp;</span>format 메서드로 문자열 정렬하기</a></span></li><li><span><a href="#숫자-개수-맞추기" data-toc-modified-id="숫자-개수-맞추기-2.13"><span class="toc-item-num">2.13&nbsp;&nbsp;</span>숫자 개수 맞추기</a></span></li></ul></li><li><span><a href="#채우기와-정렬을-조합해서-사용하기" data-toc-modified-id="채우기와-정렬을-조합해서-사용하기-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>채우기와 정렬을 조합해서 사용하기</a></span></li><li><span><a href="#퀴즈" data-toc-modified-id="퀴즈-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>퀴즈</a></span><ul class="toc-item"><li><span><a href="#1번-문제" data-toc-modified-id="1번-문제-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>1번 문제</a></span><ul class="toc-item"><li><span><a href="#정답" data-toc-modified-id="정답-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>정답</a></span></li></ul></li></ul></li></ul></div>

## 문자열 조작하기

### 문자열 바꾸기

* **replace('바꿀문자열', '새문자열')**

문자열 자체는 변경하지 않으며 바뀐 결과를 반환한다.

In [1]:
'Hello, world!'.replace('world', 'Python')

'Hello, Python!'

바뀐 결과를 유지하고 싶다면 변수에 할당해주면 된다.

In [2]:
s = 'Hello, world!'
s = s.replace('world!', 'Python')
s

'Hello, Python'

### 문자 바꾸기

translate는 문자열 안의 문자를 다른 문자로 바꾼다.

1. 먼저 str.maketrans('바꿀문자', '새문자')로 변환 테이블을 만든다.
2. 그 다음에 translate(테이블)을 사용하면 문자를 바꾼 뒤 결과를 반환한다.

In [3]:
table = str.maketrans('aeiou', '12345')
'apple'.translate(table)

'1ppl2'

### 문자열 분리하기

* split()

split()은 공백을 기준으로 문자열을 분리하여 리스트로 만든다.

In [4]:
'apple pear grape pineapple orange'.split()

['apple', 'pear', 'grape', 'pineapple', 'orange']

**split('기준문자열')** 과 같이 기준 문자열을 지정하면 기준 문자열로 문자열을 분리한다. 즉, 문자열에서 각 단어가 ,(콤마)와 공백으로 구분되어 있을 때 ', '으로 문자열을 분리하면 단어만 리스트로 만든다.

In [5]:
'apple, pear, grape, pineapple, orange'.split(', ')

['apple', 'pear', 'grape', 'pineapple', 'orange']

### 구분자 문자열과 문자열 리스트 연결하기

`join(리스트)`는 구분자 문자열과 문자열 리스트의 요소를 연결하여 문자열로 만든다.

공백 ' '에 join을 사용해서 각 문자열 사이에 공백이 들어가도록 만든다.

In [6]:
' '.join(['apple', 'pear', 'grape', 'pineapple', 'orange'])

'apple pear grape pineapple orange'

마이너스 '-'에 join을 사용하기

In [7]:
'-'.join(['apple', 'pear', 'grape', 'pineapple', 'orange'])

'apple-pear-grape-pineapple-orange'

### 소문자를 대문자로 바꾸기

`upper()`는 문자열의 문자를 모두 대문자로 바꾼다. 만약 문자열 안에 대문자가 있다면 그대로 유지된다.

In [8]:
'python'.upper()

'PYTHON'

### 대문자를 소문자로 바꾸기

`lower()`는 문자열의 문자를 모두 소문자로 부꾼다. 만약 문자열 안에 소문자가 있다면 그대로 유지된다.

In [9]:
'PYTHON'.lower()

'python'

### 왼쪽 공백 삭제하기

문자열을 사용하다 보면 공백을 삭제해야 할 경우가 생긴다. 이때는 lstrip, rstrip, strip 메서드를 사용한다.

`lstrip()`은 문자열에서 왼쪽에 있는 연속된 모든 공백을 삭제한다.(l은 왼쪽(left)을 의미).

In [10]:
'   Python   '.lstrip()

'Python   '

### 오른쪽 공백 삭제하기

`rstrip()`은 문자열에서 오른쪽에 있는 연속된 모든 공백을 삭제한다.(r은 오른쪽(right)을 의미).

In [11]:
'   Python   '.rstrip()

'   Python'

### 양쪽 공백 삭제하기

`strip()`은 문자열에서 양쪽에 있는 연속된 모든 공백을 삭제한다.

In [12]:
'   Python   '.strip()

'Python'

### 왼쪽의 특정 문자 삭제하기

`lstrip('삭제할문자들')`과 같이 삭제할 문자들을 문자열 형태로 넣어주면 문자열 왼쪽에 있는 해당 문자를 삭제한다.

다음은 문자열 왼쪽의 ,(콤마)와 .(점)을 삭제한다. 단, 여기서는 공백을 넣지 않았으므로 공백은 그대로 둔다.

In [13]:
', python.'.lstrip(',.')

' python.'

### 오른쪽 특정 문자 삭제하기

`rstrip('삭제할문자들')`과 같이 삭제할 문자들을 문자열 형태로 넣어주면 문자열 오른쪽에 있는 해당 문자를 삭제한다.

In [14]:
', python.'.rstrip(',.')

', python'

### 양쪽의 특정 문자 삭제하기

`strip('삭제할문자들')`과 같이 삭제할 문자들을 문자열 형태로 넣어주면 문자열 양쪽에 있는 해당 문자를 삭제한다.

In [15]:
', python.'.strip(',.')

' python'

> 구두점을 간단하게 삭제하기

string 모듈의 punctuation에는 모든 구두점이 들어있다. 다음과 같이 strip 메서드에 string.punctuation을 넣으면 문자열 양쪽의 모든 구두점을 간단하게 삭제할 수 있다.

In [16]:
import string

', python.'.strip(string.punctuation)

' python'

In [17]:
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

만약 공백까지 삭제하고 싶다면?

string.punctuation에 공백' '을 연결해서 넣어주면 된다.

In [18]:
', python.'.strip(string.punctuation + ' ')

'python'

물론 메서드 체이닝을 활용해도 된다.

In [19]:
', python.'.strip(string.punctuation).strip()

'python'

### 문자열을 왼쪽 정렬하기

문자열에 공백을 넣어서 원하는 위치에 정렬하기

`ljust(길이)`는 문자열을 지정된 길이로 만든 뒤 왼쪽으로 정렬하며 남는 공간을 공백으로 채운다.(l은 왼쪽(left)을 의미)

In [20]:
'python'.ljust(10)

'python    '

### 문자열을 오른쪽 정렬하기

`rjust(길이)`는 문자열을 지정된 길이로 만든 뒤 오른쪽으로 정렬하여 남는 공간을 공백으로 채운다.(r은 오른쪽(right)를 의미)

In [21]:
'python'.rjust(10)

'    python'

### 문자열을 가운데 정렬하기

`center(길이)`는 문자열을 지정된 길이로 만든 뒤 가운데로 정렬하며 남는 공간을 공백으로 채운다.

In [24]:
'python'.center(10)

'  python  '

만약 가운데로 정렬했을 때 전체 길이와 남는 공간이 모두 홀수가 된다면 왼쪽에 공백이 한 칸 더 들어간다.

In [25]:
'python'.center(11)

'   python  '

### 메서드 체이닝

문자열 메서드는 처리한 결과를 반환하도록 만들어져 있다. 따라서 메서드를 계속 연결해서 호출하는 메서드 체이닝이 가능하다. 메서드 체이닝은 메서드를 줄줄이 연결한다고 해서 메서드 체이닝(method chaining)이라 부른다.

In [26]:
'python'.rjust(10).upper()

'    PYTHON'

### 문자열 왼쪽에 0 채우기

`zfill(길이)`는 지정된 길이에 맞춰서 문자열의 왼쪽에 0을 채운다. (zero fill을 의미). 단 문자열의 길이보다 지정된 길이가 작다면 아무것도 채우지 않는다.

보통 zfill은 숫자를 일정 자릿수로 맞추고 앞자리는 0으로 채울 때 사용한다.

In [27]:
'35'.zfill(4)        # 숫자 앞에 0을 채움

'0035'

In [28]:
'3.5'.zfill(6)       # 숫자 앞에 0을 채움

'0003.5'

In [29]:
'hello'.zfill(10)    # 문자열 앞에 0을 채울 수도 있음

'00000hello'

### 문자열 위치 찾기

`find('찾을문자열')`은 문자열에서 특정 문자열을 찾아서 인덱스를 반환하고, 문자열이 없으면 -1을 반환한다. 

find는 왼쪽에서부터 문자열을 찾는데, 같은 문자열이 여러 개일 경우 처음 찾은 문자열의 인덱스를 반환한다.

In [30]:
'apple pineapple'.find('pl')

2

In [31]:
'apple pineapple'.find('xy')

-1

### 오른쪽에서부터 문자열 위치 찾기

`rfind('찾을문자열')`은 오른쪽에서부터 특정 문자열을 찾아서 인덱스를 반환하고, 문자열이 없으면 -1을 반환한다.(r은 오른쪽(right)을 의미).

In [33]:
'apple pineapple'.rfind('pl')

12

In [34]:
'apple pineapple'.rfind('xy')

-1

### 문자열 위치 찾기

find, rfind 이외에도 index, rindex로 문자열의 위치를 찾을 수 있다.

**index('찾을문자열')** 은 왼쪽에서부터 특정 문자열을 찾아서 인덱스를 반환한다.

단, 문자열이 없으면 에러를 발생시킨다. index도 문자열이 여러 개일 경우 처음 찾은 문자열의 인덱스를 반환한다.

In [35]:
'apple pineapple'.index('pl')

2

### 오른쪽에서부터 문자열 위치 찾기

`rindex('찾을문자열')`은 오른쪽에서부터 특정 문자열을 찾아서 인덱스를 반환한다. (r은 오른쪽(right)을 의미).

In [36]:
'apple pineapple'.rindex('pl')

12

### 문자열 개수 세기

* count('문자열')

count('문자열')은 현재 문자열에서 특정 문자열이 몇 번 나오는지 알아낸다.

In [37]:
'apple pineapple'.count('pl')

2

## 문자열 서식 지정자와 포매팅 사용하기

서식 지정자(format specifier)로 문자열을 만드는 방법과 format 메서드로 문자열을 만드는 문자열 포매팅(string formatting)에 대해 알아보자.

### 서식 지정자로 문자열 넣기

서식 지정자(format specifier)로 문자열 중간에 다른 문자열을 넣어보자.

* '%s' % '문자열'

In [38]:
'I am %s.' % 'james'

'I am james.'

서식 지정자는 %로 시작하고 자료형을 뜻하는 문자가 붙는다. %s는 문자열이라는 뜻이며 string의 s이다.('I am %s.'과 'james' 사이의 %는 따옴표로 묶지 않고 그대로 입력해야 한다).

문자열을 바로 지정하지 않고 변수를 지정할 수도 있다.

In [39]:
name = 'maria'
'I am %s.' % name

'I am maria.'

### 서식 지정자로 숫자 넣기

숫자를 넣어보자

* '%d' % 숫자

In [40]:
'I am %d years old.' % 20

'I am 20 years old.'

숫자는 %d를 넣고 % 뒤에 숫자를 지정하면 된다. %d는 10진 정수 decimal integer의 d이다.

### 서식 지정자로 소수점 표현하기

숫자 중에 소수점으로 된 실수를 넣어야 할 경우도 있다.

* '%f' % 숫자

In [41]:
'%f' % 2.3

'2.300000'

실수를 넣을 때는 %f를 사용하며 고정 소수점 fixed point의 f이다. %f는 기본적으로 소수점 이하 6자리까지 표시하므로 2.3은 2.300000으로 표시된다.

소수점 이하 자릿수를 지정하고 싶다면 다음과 같이 f 앞에 .(점)과 자릿수를 지정해주면 된다.

* '%.자릿수f' % 숫자

In [42]:
'%.2f' % 2.3

'2.30'

In [43]:
'%.3f' % 2.3

'2.300'

### 서식 지정자로 문자열 정리하기

% 뒤에 숫자를 붙이면 문자열을 지정된 길이로 만든 뒤 오른쪽으로 정렬하고 남는 공간을 공백으로 채운다.

* %길이s % '문자열'

In [44]:
'%10s' % 'python'

'    python'

%10s는 문자열의 길이를 10으로 만든 뒤 지정된 문자열을 넣고 오른쪽으로 정렬한다.. 따라서 문자열 'python'은 길이가 6이므로 왼쪽 공간을 공백 4칸으로 채운다.

> 자릿수가 다른 숫자 출력하기

문자열 오른쪽 정렬은 자릿수가 다른 숫자를 출력할 때 유용하다. %d와 %f도 숫자와 조합하여 오른쪽으로 정렬할 수 있다.

**%길이d**

In [45]:
'%10d' % 150

'       150'

In [46]:
'%10d' % 15000

'     15000'

실수는 다음과 같이 .(점) 앞에 정렬할 길이를 지정하고, 점 뒤에 소수점 이하 지릿수를 지정한다.

**%길이.자릿수f**

In [47]:
'%10.2f' % 2.3

'      2.30'

In [48]:
'%10.2f' % 2000.3

'   2000.30'

왼쪽 정렬은 문자열 길이에 -를 붙여주면 된다!

* %-길이d

In [49]:
'%-10s' % 'python'

'python    '

### 서식 지정자로 문자열 안에 값 여러 개 넣기

문자열 안에 값을 두 개 이상 넣으려면 %를 붙이고, 괄호 안에 값(변수)을 콤마로 구분해서 넣어주면 된다.

특히 값을 괄호로 묶지 않으면 에러가 발생하므로 주의해야 한다.

* '%d %s' % (숫자, '문자열')

In [50]:
'Today is %d %s.' % (3, 'April')

'Today is 3 April.'

서식 지정자가 여러 개면 괄호 안의 값(변수) 개수도 서식 지정자 개수와 똑같이 맞춰주어야 한다.

지금까지 서식 지정자 사이를 공백으로 띄웠다. 만약 서식 지정자를 서로 붙이면 결과도 붙어서 나오므로 주의해야 한다.

In [51]:
'Today is %d%s.' % (3, 'April')

'Today is 3April.'

###  format 메서드 사용하기

파이썬은 문자열을 만들 때 서식 지정자 방식보다 더 간단한 문자열 포매팅(string formatting)을 제공한다.

문자열 포매팅은 { }(중괄호) 안에 포매팅을 지정하고 format 메서드로 값을 넣는다.

* '{인덱스}'.format(값)

In [52]:
'Hello, {0}'.format('world!')

'Hello, world!'

In [53]:
'Hello, {0}'.format(100)

'Hello, 100'

### format 메서드로 값을 여러 개 넣기

인덱스의 순서와 format에 지정된 값의 순서에 주목하자

In [54]:
'Hello, {0} {2} {1}'.format('Python', 'Script', 3.6)

'Hello, Python 3.6 Script'

{ }에 인덱스를 지정한 뒤 format에는 인덱스가 증가하는 순서대로 값을 넣으면 된다. 여기서는 {0} {2} {1}처럼 인덱스 순서를 섞었고 format에는 'Python', 'Script', 3.6 순으로 넣었다. 따라서 문자열에는 인덱스 숫자에 해당하는 값이 들어가서 'Hello, Python 3.6 Script'가 나온다.

### format 메서드로 같은 값을 여러 개 넣기

특히 같은 인덱스가 지정된 { }를 여러 개 넣으면 같은 값이 여러 개 들어간다.

In [55]:
'{0} {0} {1} {1}'.format('Python', 'Script')

'Python Python Script Script'

### format 메서드에서 인덱스 생략하기

만약 { }에서 인덱스를 생략하면 format에 지정한 순서대로 값이 들어간다.

In [56]:
'Hello, {} {} {}'.format('Python', 'Script', 3.6)

'Hello, Python Script 3.6'

### format 메서드에서 인덱스 대신 이름 지정하기

{ }에 인덱스 대신 이름을 지정할 수도 있다.

In [57]:
'Hello, {language} {version}'.format(language='Python', version=3.6)

'Hello, Python 3.6'

### 문자열 포매팅에 변수를 그대로 사용하기

변수에 값을 넣고 { }에 변수 이름을 지정하면 된다. 이때는 문자열 앞에 포매팅(formatting)이라는 뜻으로 f를 붙인다.

In [58]:
language = 'Python'
version = 3.6
f'Hello, {language} {version}'

'Hello, Python 3.6'

> 중괄호 출력하기

{ } 중괄호 자체를 출력할 때는 {{,}}처럼 중괄호를 두 번 사용하면 된다.

In [59]:
'{{ {0} }}'.format('Python')

'{ Python }'

### format 메서드로 문자열 정렬하기

인덱스 뒤에 :(콜론)을 붙이고 정렬할 방향과 길이를 지정해주면 된다.

* '{인덱스:<길이}'.format(값)

In [61]:
'{0:<10}'.format('python')

'python    '

\>을 넣어서 오른쪽을 가리키도록 만들면 문자열을 지정된 길이로 만든 뒤 오른쪽으로 정렬하고 남는 공간을 공백으로 채운다.

* '{인덱스:>길이}'.format(값)

In [62]:
'{0:>10}'.format('python')

'    python'

참고로 인덱스를 사용하지 않는다면 :(콜론)과 정렬 방법만 지정해도 된다.

In [63]:
'{:>10}'.format('python')

'    python'

### 숫자 개수 맞추기

* '%0개수d' % 숫자
* '{인덱스:0개수d'}'.format(숫자)

In [64]:
'%03d' % 1

'001'

In [65]:
'{0:03d}'.format(35)

'035'

In [66]:
'{:03d}'.format(35)

'035'

실수로 숫자 개수 맞추기

* '%0개수.자릿수f' % 숫자
* '{인덱스:0개수.자릿수f'}.format(숫자)

In [67]:
'%08.2f' % 3.6

'00003.60'

In [68]:
'{0:08.2f}'.format(150.37)

'00150.37'

In [69]:
'{:08.2f}'.format(150.37)

'00150.37'

여기서 주의할 점은 '%08.2f' % 3.6을 출력했을 때 '00003.60'이 나온다는 점이다. 잘 보면 3 앞에 0이 7개가 아니라 4개만 붙어 있다. 실수는 숫자 개수에 정수 부분, 소수점, 소수점 이하 자릿수가 모두 포함된다. 따라서 '00003' 5개, '.' 1개, '60' 2개 총 8개이다.

* 정수 부분, 소수점, 소수점 이하 자릿수 모두 포함!

## 채우기와 정렬을 조합해서 사용하기

문자열 포매팅은 채우기와 정렬을 조합해서 사용할 수 있음.

다음과 같이 { }에 인덱스, 채우기, 정렬, 길이, 자릿수, 자료형 순으로 지정해준다.

* '{인덱스:[[채우기]정렬][길이][.자릿수][자료형]}'

In [70]:
'{0:0<10}'.format(15)    # 길이 10, 왼쪽으로 정렬하고 남는 공간은 0으로 채움

'1500000000'

In [71]:
'{0:0>10}'.format(15)    # 길이 10, 오른쪽으로 정렬하고 남는 공간은 0으로 채움

'0000000015'

실수로 만들기

In [72]:
'{0:0>10.2f}'.format(15)    # 길이 10, 오른쪽으로 정렬하고 소수점 자릿수는 2자리

'0000015.00'

특히 채우기 부분에 0이 아닌 다른 문자를 넣어도 된다. 공백을 넣어도 되고, 문자를 넣어도 된다. 만약 채우기 부분을 생략하면 공백이 들어간다.

In [73]:
'{0: >10}'.format(15)    # 남는 공간을 공백으로 채움

'        15'

In [74]:
'{0:>10}'.format(15)     # 채우기 부분을 생략하면 공백이 들어감

'        15'

In [75]:
'{0:x>10}'.format(15)    # 남는 공간을 문자 x로 채움

'xxxxxxxx15'

> 금액에서 천단위로 콤마 넣기

숫자 중에서 금액은 천단위로 ,(콤마)를 넣는다. 파이썬에서는 간단하게 천단위로 콤마를 넣을 수 있다.

먼저 format 내장 함수를 사용하는 방법이다. 다음과 같이 format 함수에 숫자와 ','를 넣어준다.

**format(숫자, ',')**

In [77]:
format(1493500, ',')

'1,493,500'

format 함수는 서식 지정자와 함께 사용할 수 있다. 다음은 콤마를 넣은 숫자를 오른쪽 정렬한다.

In [78]:
'%20s' % format(1493500, ',')    # 길이 20, 오른쪽으로 정렬

'           1,493,500'

포매팅에서 콤마를 넣으려면 다음과 같이 :(콜론)뒤에 ,(콤마)를 지정하면 된다.

In [79]:
'{0:,}'.format(1493500)

'1,493,500'

만약 정렬을 하고 싶다면 정렬 방향과 길이 뒤에 콤마를 지정해준다.

In [80]:
'{0:>20,}'.format(1493500)     # 길이 20, 오른쪽으로 정렬

'           1,493,500'

In [81]:
'{0:0>20,}'.format(1493500)    # 길이 20, 오른쪽으로 정렬하고 남는 공간은 0으로 채움

'000000000001,493,500'

## 퀴즈

### 1번 문제

In [84]:
path = 'C:\\Users\\dojang\\AppData\\Local\\Programs\\Python\\Python36-32\\python.exe'

l = path.split('\\')
filename = l[-1]

print(filename)

python.exe


#### 정답

In [89]:
x = path.split('\\')
x.reverse()
filename = x[0]

In [86]:
filename = path[path.rfind('\\') + 1:]
filename

'python.exe'

In [88]:
path.rfind('\\') + 1

58

> raw 문자열 사용하기

문자열 앞에 r 또는 R을 붙이면 raw 문자열이 된다. 이 raw 문자열은 이스케이프 시퀀스를 그대로 저장할 때 사용한다. 즉, 다음과 같이 \를 \\\로 두 번 쓰지 않고 한 번만 써도 된다.

In [90]:
print(r'C:\Users\dojang\AppData\Local\Programs\Python\Python36-32\python.exe')

C:\Users\dojang\AppData\Local\Programs\Python\Python36-32\python.exe


raw는 가공되지 않고 있는 그대로라는 뜻이다. 따라서 이스케이프 시퀀스를 문자 그대로 표현한다. 다음과 같이 raw 문자열에 제어 문자를 입력해보면 제어 문자가 동작하지 않는 것을 볼 수 있다.

In [91]:
print(r'1\n2\n3\n')

1\n2\n3\n
